src/controllers/midi/midisourceclock.cpp

```diff
@@ -0,0 +1,117 @@
+#include "controllers/midi/midimessage.h"
+#include "controllers/midi/midisourceclock.h"
+#include "util/math.h"
+
+bool MidiSourceClock::handleMessage(unsigned char status) {
+    switch (status) {
+    case MIDI_START:
+        start();
+        return true;
+    case MIDI_STOP:
+        stop();
+        return true;
+    case MIDI_TIMING_CLK:
+        tick();
+        return true;
+    default:
+        return false;
+    }
+}
+
+void MidiSourceClock::start() {
+    m_bRunning = true;
+    m_iFilled = 0;
+    m_iRingBufferPos = 0;
+}
+
+void MidiSourceClock::stop() {
+    m_bRunning = false;
+}
+
+void MidiSourceClock::tick() {
+    // Update the ring buffer and calculate new bpm.  Update the last beat time
+    // if we are on a beat.
+
+    if (!m_bRunning) {
+        qDebug() << "MidiSourceClock: Got clock tick but not started, starting now.";
+        start();
+    }
+
+    // Ringbuffer filling.
+    // TODO(owen): We should have a ringbuffer convenience class.
+    const qint64 lastTickTime = m_pClock->now();
+    m_iTickRingBuffer[m_iRingBufferPos] = lastTickTime;
+    m_iRingBufferPos = (m_iRingBufferPos + 1) % kRingBufferSize;
+    if (m_iFilled < kRingBufferSize) {
+        ++m_iFilled;
+    }
+
+    // If this tick is a beat mark, record it, even if we have very few samples.
+    if (m_iRingBufferPos % kPulsesPerQuarter == 0) {
+        m_iLastBeatTime = lastTickTime;
+    }
```

Comment:
> The midi standard does not communicate measures.

Yes, I just wonder if there is a quasi standard for measures since all connected devices have to map their measures to the midi clock ticks. 

According to https://en.wikipedia.org/wiki/Tempo Quaters per minute is the same as beats per minute. Quarter would be a new term in the Mixxx syncing code so I prefer to stick at "beat".